File: src/main/java/com/test/domain/UserProfile.java
package com.test.domain;

import com.test.util.MessageComparator;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by alvarovilaplanagarcia on 21/04/15.
 */
public class UserProfile {
    private final String user;
    private final List<Message> messages = new LinkedList<Message>();
    private final List<String> follow = new LinkedList<String>();

    public UserProfile(final String user) {
        this.user = user;
    }

    public List<Message> addMessage(final Message message) {
        messages.add(message);
        return messages;
    }

    public List<Message> getMessages() {
        Collections.sort(messages, new MessageComparator());
        return Collections.unmodifiableList(messages);
    }

    public List<String> addFollow(final String user) {
        follow.add(user);
        return follow;
    }

    public List<String> getFollow() {
        return Collections.unmodifiableList(follow);
    }

    public String getUser() {
        return user;
    }

    public UserProfile copy() {
        UserProfile userProfileCopied = new UserProfile(this.user);
        for (Message message : messages) {
            userProfileCopied.addMessage(message);
        }

        for (String user : follow) {
            userProfileCopied.addFollow(user);
        }
        return userProfileCopied;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof UserProfile)) return false;

        UserProfile that = (UserProfile) o;

        if (!follow.equals(that.follow)) return false;
        if (!messages.equals(that.messages)) return false;
        if (!user.equals(that.user)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = user.hashCode();
        result = 31 * result + messages.hashCode();
        result = 31 * result + follow.hashCode();
        return result;
    }
}

File: scala/src/main/scala/com/test/domain/UserProfile.scala
package com.test.domain

import org.joda.time.DateTime

case class Message(user: String, message: String, created: DateTime)

case class UserProfile(messages: List[Message] = List.empty[Message], follow: List[String] = List.empty[String]) {

  def addMessage(message: Message) = this.copy(messages = this.messages :+ message)

  def addFollow(user: String) = this.copy(follow = this.follow :+ user)

  def messagesOrdered(): List[Message] = messages.sortWith((m1, m2) => m1.created.isAfter(m2.created))
}

trait ProfileRepository {
  def post(user: String, message: String, created: DateTime): UserProfile

  def reading(user: String): List[Message]

  def follows(userFollows: String, userFollowed: String): Option[UserProfile]

  def wall(use: String): List[Message]
}

class UserProfileRepository extends ProfileRepository {
  val profiles = scala.collection.mutable.Map.empty[String, UserProfile]

  override def post(user: String, message: String, created: DateTime): UserProfile =
    profiles.get(user) match {
      case Some(up) =>
        val profileUpdated = up.addMessage(Message(user, message, created))
        profiles += (user -> profileUpdated)
        profileUpdated
      case None =>
        val newProfile = UserProfile(List(Message(user, message, created)))
        profiles += (user -> newProfile)
        newProfile
    }

  override def reading(user: String): List[Message] =
    profiles.get(user) match {
      case Some(up) => up.messagesOrdered()
      case None => List()
    }

  override def follows(userFollows: String, userFollowed: String): Option[UserProfile] =
    profiles.get(userFollows) match {
      case Some(up) =>
        val profileUpdated = up.addFollow(userFollowed)
        profiles += (userFollows -> profileUpdated)
        Some(profileUpdated)
      case None => None
    }

  override def wall(user: String): List[Message] =
    profiles.get(user) match {
      case Some(up) =>
        val followMessages: List[Message] = up.follow.flatMap(fu => profiles.get(fu).map(_.messagesOrdered())).flatten.toList
        val allMessages = up.messagesOrdered() ++ followMessages
        allMessages.sortWith((m1, m2) => m1.created.isAfter(m2.created))
      case None => List.empty[Message]
    }
}
